### PR TITLE
ADD: Zhongkai's code

### PR DESCRIPTION
**Pull request by marques-bruno**
Hello,

I've taken the code from Zhongkai's gitlab repository and added it to the SoftRobots plugin. Basically, this adds a new component that retrieves & processes point clouds coming from a RealSense camera. I didn't look into the details but it seems to be standard segmenting, smoothing, clustering, outlier removal etc for 3D point clouds.

It's deactivated by default, and can be enabled through cmake by setting the SOFTROBOTS_WITH_CAMERA option to ON in cmake. Some small installation documentation is also added in the readme.md, explaining how to install the 2 dependencies for the plugin:
- the PCL library
- The kernel module for realsense & its associated SDK.

I don't know if master should be the target branch to merge this PR. if you prefer to have this on a separate branch I can change that, but it'll be at the cost of maintainability...

### CODE REVIEW

**Comment by marques-bruno** [comment]
@damienmarchal  do you know what this `libSceneCreator.so` is? When running tests on SoftRobots: `"error while loading shared libraries: libSceneCreator.so: cannot open shared object file: No such file or directory"`

**Comment by damienmarchal** [comment]
SceneCreator is a Sofa plugin that do more or less what simpleapi do but in the wrong way ;)

**Comment by marques-bruno** [comment]
[ci-build]

File: src/SoftRobots/component/camera/PointCloudProcessing.cpp [hunk 1]
```diff
@@ -0,0 +1,537 @@
...
+
+void PointCloudProcessing::update()
+{
+    pcl::PointCloud<pcl::PointXYZRGB>::Ptr filteredpoints = pclcamera.updatePcl();
+
+    //user-defined
+    //The following vector and matrix are the calibration matrix between the camera coordinate system and the robot coordinate system.
+    double x0 = -14.18895169; //unit is millimeter
+    double y0 = 35.32077316; //unit is millimeter
+    double z0 = 578.52358217; //unit is millimeter
```
Comment (EulalieCoevoet) [inline src/SoftRobots/component/camera/PointCloudProcessing.cpp]: I guess it's only supposed to work with a specific configuration...
Can we fix that before merging...? Add a data maybe?

File: src/SoftRobots/component/camera/PointCloudProcessing.cpp [hunk 2]
```diff
@@ -0,0 +1,537 @@
...
+    //it is used for merging process mentioned in the beginning. If cluster has less points than was set through setMinClusterSize method, then it will be merged with the nearest neighbour.
+    reg.setMinClusterSize (100); //set the size of the class
+
+    //Here is the place where the algorithm is launched. It will return the array of clusters when the segmentation process will be over.
+    std::vector <pcl::PointIndices> clusters;
+    reg.extract (clusters);
+
+    std::cout << "Number of clusters is equal to " << clusters.size () << std::endl;
+    std::cout << "First cluster has " << clusters[0].indices.size () << " points." << endl;
+
```
Comment (EulalieCoevoet) [inline src/SoftRobots/component/camera/PointCloudProcessing.cpp]: I feel like this contribution could be cleaned during a Friday coding session...? maybe

File: src/SoftRobots/component/camera/PointCloudProcessing.cpp [hunk 3]
```diff
@@ -0,0 +1,537 @@
...
+    d_goalPositions(initData(&d_goalPositions, "goalPositions", "Position of the contact forces.")),
+    d_normalDirections(initData(&d_normalDirections, "normalDirection", "The normal directions of effectors")),
+    d_contactLocations(initData(&d_contactLocations, "contactLocations", "The index of contact locations"))
+{
+    pclcamera.initCamera();
+}
+
+
+void PointCloudProcessing::update()
+{
```
Comment (EulalieCoevoet) [inline src/SoftRobots/component/camera/PointCloudProcessing.cpp]: This update function could use some refactoring...

**Comment by damienmarchal** [comment]
Total agreement with Eulalie.

**Comment by marques-bruno** [comment]
Total agreement with both of you, but I wanted to limit my involvement on this for now (in fact I didn't even look at the code yet).
I think Friday's coding session will be perfect for this task though. I'll ask @zhongkaidefrost to help me with the specifics. I guess the same comments apply to the SoftRobots.Inverse merge request.

**Comment by marques-bruno** [comment]
@zhongkaidefrost, I've refactored the code to make it more readable (update is cut into multiple subfunctions that I've identified: **image segmentation**, **computation of the effectors normal directions**, **computation of the desired 3D position for the robot** and finally **the location of the contacts**).

Please look at the code and tell me if the way I've partitioned the code sounds allright to you.
Also, I've exposed a certain number of parameters (from the different PCL algorithms used in the code) as data fields in SOFA, so that they can be tuned without digging in & recompiling the code. Some might not actually be meaningful to expose, and some that I didn't expose might be... Here too, please tell me if it makes sense to you. I wait for your approval (and of course @EulalieCoevoet & @damienmarchal 's approvals too... :) )

**Review by EulalieCoevoet — APPROVED** [review]

**Pull request merged**